Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1380 @@
+
+**Pros:**
+* Parser generators report ambiguous grammars to the user, creating greater trust in the parser's output and saving from debugging headaches.
+* Bottom-up parsers can handle left recursion, while top-down parsers can't.
+* The performance will be more predictable, as Megaparsec can cause exponential time complexity with the use of `try`. In addition, Happy's "static" generated LALR algorithm has better known complexity constraints.
+* Dealing with operator precedence and associativity requires less labor. In Megaparsec, there are standard tools like [`makeExprParser`](https://hackage.haskell.org/package/parser-combinators-1.3.0/docs/Control-Monad-Combinators-Expr.html#v:makeExprParser) to deal with them, but once you need something that is non-standard, you'll need to write more non-trivial parser rules by hand.
+
+**Cons:**
+* Less idiomatic, as you need to write Happy grammars instead of Haskell code.
+* Less flexible since you can't change Happy’s algorithm. Meanwhile, parser combinators enable you to easily create your own functions.
+
+## How Happy's parsers work
+
+An in-depth explanation of how an LR(1) parser works is well beyond the scope of this tutorial. For more info on this, we recommend you read our article on [how to implement an LR(1) Parser](https://serokell.io/blog/how-to-implement-lr1-parser).
+We will, however, introduce a few basic concepts required to understand this article better.
+
+### Formal grammar 
+
+When talking about parsing, we generally imply we're parsing some language, a formal language in the case of programming languages.
+But what is a language?
+
+Formal language theory offers a very simple answer: a language is a (likely infinite) set of all grammatically correct sentences.
+Now, the word "sentence" is used in a very general sense here – a complete MiniML program would still constitute a sentence.
+
+Well, but how do you describe an infinite set of sentences?
+You do that by specifying how how to build any arbitrary sentence.
+
+This description is called the formal grammar of the language.
+The usual approach uses string substitutions to construct a sentence.
+The process of constructing a sentence itself is called the derivation process, which is represented as a sequence of substitutions called a derivation chain.
+We can refer to either as **derivation**.
+
+We start with a special symbol called the start non-terminal. Then, we find and replace symbols according to our substitution rules. Once we have replaced everything we could have replaced, we stop.
+
+The symbols to be replaced are called **non-terminals**, and the symbols they are ultimately replaced by are called **terminals**. Non-terminals denote things that do not occur in the language itself. Terminals are usually the lexical items of the language – you can think of them being the "words" of the language. They are called so because derivation terminates on them.
```

Comment:
A couple small nitpicks here:

> Non-terminals denote things that do not occur in the language itself

This is not exactly correct. The symbols we choose for non-terminals do not appear in the language (in some general sense), but the things they denote obviously do.

> Terminals are usually the lexical items of the language

Again, more correct to say terminals represent lexical items of the language; a lexeme might be "x", but the corresponding terminal is not necessarily "x", it might be "identifier", for example.